share/doc/wake/logging.md

```diff
@@ -0,0 +1,152 @@
+----------------------------|----------------------------------|
+|  logger                   | `LogLevel`                       |
+|  log level                | `LogLevel.name`                  |
+
+## Using & Creating Loggers
+
+`wake` has a number of built in loggers in which all start with `log...`.
+You can also make your own logger using `mkLogLevel`.
+You can use this to customize the color you want it to be shown as (if at all)
+and give more fine-grained control over which wake output stream it should go to.
+
+
+| wake `def`        | Color/Intensity | log level |
+|-------------------|-----------------|-----------|
+| logDebug          |     Blue        | "debug"   |
+| logInfo           |     Dim         | "info"    |
+| logEcho           |      N/A        | "echo"    |
+| logReport         |    Magenta      | "report"  |
+| logWarning        |    Yellow       | "warning" |
+| logError          |     Red         | "error"   |
+| logNever          |     N/A         | "null"    |
+| mkLogLevel "foo" Green |  Green     | "foo"     |
+| mkLogLevel "foo"  Bright Green | Bright Green   | "foo"      |
+
+## Directing Logger Output
+
+Based on command line flags and its log level,
+each logger's output will be directed to zero or more of wake's outputs streams.
+
+For `wake`'s own stdout, the logger's color & intensity are used to apply
+additional formatting to the output before displaying it.
+
+For understanding which streams will appear on wake's stdout,
+the following table should be read top to bottom considering the command line arguments you've passed to wake.
+If you match a line in the table, stop!
+This is what will appear on the console as you watch wake execute,
+i.e. what is sent to ``wake`'s stdout:
+
+| log level           | "debug"  | "info"  | "echo"  | "report"  | "warning"  | "error"  | "null"  |  "foo" |
+|---------------------|----------|---------|---------|-----------|------------|----------|---------|--------|
+|--stdout="foo"       |          |         |         |           |            |          |         |    x   |
+|--stdout="foo,info"  |          |    x    |         |           |            |          |         |    x   |
+|--debug              |      x   |    x    |    x    |     x     |      x     |    x     |         |        |
+|--verbose            |          |    x    |    x    |     x     |      x     |    x     |         |        |
+|--quiet              |          |         |         |           |            |    x     |         |        |
+|--no-tty<sup>*</sup> |          |         |    x    |     x     |      x     |    x     |         |        |
+|(default)            |          |         |         |     x     |      x     |    x     |         |        |
+
+<sup>*</sup> `--no-tty` also prevents any coloring from being applied.
+
+Unless overridden with `--stderr=...` the *only* thing that ever appears on wake's stderr is outputs from loggers
+with log level "error" (e.g. `logError`).
```

Comment:
@JakeSiFive, @V-FEXrt: Do we make any tighter guarantees about the stderr than we do the stdout?  It's the only thing that appears once you hit the wake code, definitely, but there might be more things output by the tool itself (slow disc warnings?) that appear here.

Also, I'm still not quite clear on how both stdout and stderr naming `"error"` doesn't result in duplicated output as Wake prints it to both file descriptors; we definitely don't have to go into the details in this document, but it might be a nice thing to briefly mention, especially if the mechanism is less-than-intuitive (e.g. something like *only* Job stderr is put on Wake's stderr).

Reply:
> Do we make any tighter guarantees about the stderr than we do the stdout? 
No, stderr and stdout are not safe to depend on in any general context. The only guaranteed way to have a clean output/capture is to create a new stream (because anyone can output to the default streams) and then to redirect said stream to an fd (which may be stdout/stderr if desired)

Reply:
>  I'm still not quite clear on how both stdout and stderr naming "error" doesn't result in duplicated output as Wake prints it to both file descriptors; 

I am also not clear on this, this came out of just me reading the code so I could be wrong.

Reply:
Ok yeah I should tighten up there may be other cruft on stderr